src/coreComponents/physicsSolvers/fluidFlow/OBLSuperEngine.cpp

```diff
@@ -0,0 +1,1283 @@
+    }
+    subRegionSetMap[setName].setNumComp( numComp );
+  } );
+
+  // 2. Check composition BC (global component fraction)
+  fsManager.apply( time,
+                   domain,
+                   "ElementRegions",
+                   extrinsicMeshData::flow::globalCompFraction::key(),
+                   [&] ( FieldSpecificationBase const & fs,
+                         string const & setName,
+                         SortedArrayView< localIndex const > const &,
+                         Group & subRegion,
+                         string const & )
+  {
+    // 2.0. Check pressure and record composition bc application
+    string const & subRegionName = subRegion.getName();
+    string const & regionName = subRegion.getParent().getParent().getName();
+    integer const comp = fs.getComponent();
+
+    auto & subRegionSetMap = bcStatusMap[regionName][subRegionName];
+    if( subRegionSetMap.count( setName ) == 0 )
+    {
+      bcConsistent = false;
+      GEOSX_WARNING( GEOSX_FMT( "Pressure boundary condition not prescribed on set {}/{}/{}", regionName, subRegionName, setName ) );
+    }
+    if( comp < 0 || comp >= numComp )
+    {
+      bcConsistent = false;
+      GEOSX_WARNING( GEOSX_FMT( "Invalid component index [{}] in composition boundary condition {}", comp, fs.getName() ) );
+      return;     // can't check next part with invalid component id
+    }
+
+    ComponentMask< MAX_NC > & compMask = subRegionSetMap[setName];
+    if( compMask[comp] )
+    {
+      bcConsistent = false;
+      GEOSX_WARNING( GEOSX_FMT( "Conflicting composition[{}] boundary conditions on set {}/{}/{}", comp, regionName, subRegionName, setName ) );
+    }
+    compMask.set( comp );
+  } );
+
+  // 2.3 Check consistency between composition BC applied to sets
+  for( auto const & regionEntry : bcStatusMap )
+  {
+    for( auto const & subRegionEntry : regionEntry.second )
+    {
+      for( auto const & setEntry : subRegionEntry.second )
+      {
+        ComponentMask< MAX_NC > const & compMask = setEntry.second;
+        for( integer ic = 0; ic < numComp; ++ic )
+        {
+          if( !compMask[ic] )
+          {
+            bcConsistent = false;
+            GEOSX_WARNING( GEOSX_FMT( "Boundary condition not applied to composition[{}] on set {}/{}/{}",
+                                      ic, regionEntry.first, subRegionEntry.first, setEntry.first ) );
+          }
+        }
+      }
+    }
+  }
+
+  return bcConsistent;
+}
+
+}
+
+
+void OBLSuperEngine::applyDirichletBC( real64 const time,
+                                       real64 const dt,
+                                       DofManager const & dofManager,
+                                       DomainPartition & domain,
+                                       CRSMatrixView< real64, globalIndex const > const & localMatrix,
+                                       arrayView1d< real64 > const & localRhs ) const
```

Comment:
I agree pressure and temperature checks and applications are similar, while for compositions it's more difficult because of different primary variable choices. It looks like a good candidate for refactoring PR - I also think it's more reliable to apply the tik-tok strategy here (implementation-refactoring).